### PR TITLE
Update llvm-project to Chromium m114

### PR DESCRIPTION
**Pull request by niranjanyardi**
b/323184809

Update cobalt's llvm rep to Chromium's m114 llvm pinned at llvmorg-17-init-8029-g27f27d15

Build llvm repo files with C++20(same as upstream)

Test-On-Device: true

### CODE REVIEW

File: third_party/llvm-project/libcxx/src/algorithm.cpp [hunk 1]
```diff
@@ -11,9 +11,24 @@
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
+// The definition for std::__bit_log2 is in
```
Comment (niranjanyardi) [inline third_party/llvm-project/libcxx/src/algorithm.cpp:14]: @kaidokert  ptal

Comment (kaidokert) [reply third_party/llvm-project/libcxx/src/algorithm.cpp:14]: Why does this conflict happen though ? Why is a call site invoking a call to std::__bit_log2 if it's being compiled with c++17 settings ?

File: third_party/llvm-project/libcxx/src/algorithm.cpp [hunk 2]
```diff
@@ -11,9 +11,24 @@
...
+// The definition for std::__bit_log2 is in 
+// llvm-project/libcxx/include/__bit/bit_log2.h and is gaurded by _LIBCPP_STD_VER >= 20
+// Since we are building with C++17 we are adding _LIBCPP_STD_VER >= 20
+
+
+
 template <class Comp, class RandomAccessIterator>
 void __sort(RandomAccessIterator first, RandomAccessIterator last, Comp comp) {
+
+#if defined USE_COBALT_CUSTOMIZATIONS
```
Comment (kaidokert) [inline third_party/llvm-project/libcxx/src/algorithm.cpp:23]: Do we need this, or is the _LIBCPP_STD_VER sufficient ? Note that in the base/net update feature branch `_LIBCPP_STD_VER` issue is separately worked on, we need to be consistent about setting it. There's an open PR somewhere for it

Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/src/algorithm.cpp:23]: I wanted to explicitly say this is a cobalt customization , so someone updating llvm next time knows this while resolving the conflict.

File: docker/linux/android/Dockerfile [hunk 3]
```diff
@@ -51,6 +52,8 @@ ARG TOOLS=commandlinetools-linux-6200805_latest.zip
 ARG CMDLINE_URL=https://dl.google.com/android/repository/${TOOLS}
 ENV ANDROID_SDK_ROOT=/root/starboard-toolchains/AndroidSdk/
 
+RUN rm -rf ${ANDROID_SDK_ROOT}
+
 RUN cd /tmp \
   && mkdir -p ${ANDROID_SDK_ROOT} \
```
Comment (kaidokert) [inline docker/linux/android/Dockerfile:58]: Move the rm -rf to be between `cd` and `mkdir` here, this way we avoid another RUN command that would create an extra container layer.

**Comment by niranjanyardi** [comment]
Android error looks unrelated: 
Cause: com.google.devtools.common.metrics.stability.converter.DeserializedException: Invalid configuration for device 28201HFDD7HYXS (test type perf)

File: starboard/raspi/shared/test_filters.py [hunk 4]
```diff
@@ -39,6 +39,11 @@
 
         # Enable once verified on the platform.
         'SbMediaCanPlayMimeAndKeySystem.MinimumSupport',
+
+        # Passes locally but fails on github.b/xx
```
Comment (niranjanyardi) [inline starboard/raspi/shared/test_filters.py:43]: note these tests are flaky: https://github.com/youtube/cobalt/actions/runs/8061465385/job/22019784016?pr=2494

Comment (niranjanyardi) [reply starboard/raspi/shared/test_filters.py:43]: These tests pass locally b/327686141

File: third_party/llvm-project/libcxx/src/algorithm.cpp [hunk 5]
```diff
@@ -11,9 +11,24 @@
...
+
 template <class Comp, class RandomAccessIterator>
 void __sort(RandomAccessIterator first, RandomAccessIterator last, Comp comp) {
+
+#if defined USE_COBALT_CUSTOMIZATIONS  
+#if _LIBCPP_STD_VER >= 20
   auto depth_limit = 2 * std::__bit_log2(static_cast<size_t>(last - first));
+#else
+  // horrible fix, never submit this
+  auto depth_limit = 2 * std::log2(static_cast<size_t>(last - first));
```
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/src/algorithm.cpp]: got past this error by upgrading to c++20

**Comment by niranjanyardi** [comment]
> Android error looks unrelated: Cause: com.google.devtools.common.metrics.stability.converter.DeserializedException: Invalid configuration for device 28201HFDD7HYXS (test type perf)
> 
> [evergreen-arm-hardfp / sbversion-14_on_device_0](https://github.com/youtube/cobalt/actions/runs/8093087190/job/22115451535?pr=2484#logs) Looks like tests are passing: [screenshot.googleplex.com/6SkNSwsgkgASZtR](https://screenshot.googleplex.com/6SkNSwsgkgASZtR)
> 
> [evergreen-arm-hardfp / sbversion-15_on_device_0] ([youtube/cobalt/actions/runs/8093087190/job/22115450416?pr=2484#logs](https://github.com/youtube/cobalt/actions/runs/8093087190/job/22115450416?pr=2484#logs)) Cause: com.google.devtools.common.metrics.stability.converter.DeserializedException: Test timeout and interrupted [MH|CUSTOMER_ISSUE|TR_JOB_TIMEOUT_AND_INTERRUPTED|40107] [MobileHarnessException] Looks like test timed out . Test seems flaky : [screenshot.googleplex.com/BYioMX7ZxUyTEWg](https://screenshot.googleplex.com/BYioMX7ZxUyTEWg) . i made a dummy change and it fails there too

Latest failures:

//video/youtube/testing/maneki/cobalt/unit_tests:cobalt_android_x86_unit_tests
http://shortn/_HQFJn85rO8
I0313 10:09:54.385514 140299271599872 process_runner.py:30] DbgFsParseFile: fopen of /d/pvr/driver_stats failed (Permission denied)

File: docker/linux/android/Dockerfile (hunk 3 above)
Comment (niranjanyardi) [reply docker/linux/android/Dockerfile:59]: done

File: starboard/build/config/BUILDCONFIG.gn [hunk 6]
```diff
@@ -216,6 +217,14 @@ template("target_with_platform_configs") {
       configs += [ "//starboard/build/config:pedantic_warnings" ]
     }
     configs += [ toolchain_config_path ]
+
+    # if (sb_is_modular && current_toolchain == cobalt_toolchain) {
```
Comment (niranjanyardi) [inline starboard/build/config/BUILDCONFIG.gn:221]: this is weird, i dont see the change locally when i checkout this branch

File: starboard/raspi/shared/test_filters.py [hunk 7]
```diff
@@ -39,6 +39,11 @@
 
         # Enable once verified on the platform.
         'SbMediaCanPlayMimeAndKeySystem.MinimumSupport',
+
+        # TODO: b/327686141 - Passes locally but fails on github
+        'SbSocketAddressTypes/SbSocketGetInterfaceAddressTest.SunnyDayDestination/type_ipv6',
+        'SbSocketAddressTypes/SbSocketGetInterfaceAddressTest.SunnyDaySourceForDestination/type_ipv6',
+        'SbSocketAddressTypes/SbSocketGetInterfaceAddressTest.SunnyDaySourceNotLoopback/type_ipv6',
```
Comment (niranjanyardi) [inline starboard/raspi/shared/test_filters.py]: this whole commit has been already cherry-picked in trunk

File: docker/linux/android/Dockerfile
Comment (andrewsavage1) [inline docker/linux/android/Dockerfile:1]: Is this something that needs to be done in this PR? Or can it be done separately?

File: starboard/build/config/BUILDCONFIG.gn (hunk 6 above)
Comment (andrewsavage1) [reply starboard/build/config/BUILDCONFIG.gn:221]: Let's just make sure this doesn't end up in trunk

File: starboard/build/config/modular/BUILD.gn [hunk 8]
```diff
@@ -12,6 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+config("default_cpp_standard") {
```
Comment (andrewsavage1) [inline starboard/build/config/modular/BUILD.gn:15]: Is this something we'll eventually want to have on other platforms, or will we only ever need/want it for modular builds?

File: third_party/protobuf/BUILD.gn [hunk 9]
```diff
@@ -23,6 +23,8 @@ config("protobuf_config") {
         "-Wno-enum-compare",
         # Older version of clang don't have -Wno-enum-compare-switch
         "-Wno-unknown-warning-option",
+
+        "-Wno-deprecated-declarations",
```
Comment (andrewsavage1) [inline third_party/protobuf/BUILD.gn:27]: What is this needed for? Should there be a cleanup bug for this?

File: docker/linux/android/Dockerfile
Comment (niranjanyardi) [reply docker/linux/android/Dockerfile:1]: The builder was failing - re -running it didnt help . So i made this change. In theory : it shouldnt' affect this change

File: starboard/build/config/modular/BUILD.gn (hunk 8 above)
Comment (niranjanyardi) [reply starboard/build/config/modular/BUILD.gn:15]: this wont be the case for all platforms : think clang3-9 ,  host toolchain which build with cpp14.

File: third_party/protobuf/BUILD.gn (hunk 9 above)
Comment (niranjanyardi) [reply third_party/protobuf/BUILD.gn:27]: I ran into a compiler error when i switched to cpp20 -  but maybe that was because of a misconfiguration. ill try removing this , rerunning and seeing what happens.

File: starboard/build/config/BUILDCONFIG.gn (hunk 6 above)
Comment (niranjanyardi) [reply starboard/build/config/BUILDCONFIG.gn:221]: will do! still dont see it when i fetch the upstream change

File: third_party/protobuf/BUILD.gn (hunk 9 above)
Comment (niranjanyardi) [reply third_party/protobuf/BUILD.gn]: removed, tried a small test , saw it works locally. Added a new commit to remove it

File: starboard/build/config/modular/BUILD.gn (hunk 8 above)
Comment (andrewsavage1) [reply starboard/build/config/modular/BUILD.gn:15]: I mean will we eventually want to extract the standard from the platform's configuration and have it in a platform-agnostic spot like this? Or will we only ever want to extract it from modular builds?

Comment (niranjanyardi) [reply starboard/build/config/modular/BUILD.gn:15]: i think it'll be good to extract it from all non-third party code and keep it in a single file with multiple configs: cpp14, cpp17, cpp20 . it does cause confusion to have it part of the platform flags - the flags end up overlapping - for ex; for llvm : cpp17 came from platform config whereas cpp20 came from llvm repo.

Comment (andrewsavage1) [reply starboard/build/config/modular/BUILD.gn:15]: Okay, then let's add it to a more general location, i.e. `//starboard/build/config/BUILD.gn`

Comment (niranjanyardi) [reply starboard/build/config/modular/BUILD.gn]: good idea, done

File: starboard/build/config/BUILD.gn [hunk 10]
```diff
@@ -238,3 +238,10 @@ config("default_compiler_flags") {
     asmflags = cflags
   }
 }
+
+# TODO: b/329683364 - Add non third party cpp standard flags.
```
Comment (andrewsavage1) [inline starboard/build/config/BUILD.gn:242]: Could you rephrase this comment? I.e. "Move other flags setting the c++ standard from each platform's platform_configuration to this config."

File: starboard/build/config/BUILD.gn [hunk 11]
```diff
@@ -239,7 +239,8 @@ config("default_compiler_flags") {
   }
 }
 
-# TODO: b/329683364 - Add non third party cpp standard flags.
+# TODO: b/329683364 - Move other flags setting the c++ standard from each
+# platform's platform_configuration to this config..
```
Comment (andrewsavage1) [inline starboard/build/config/BUILD.gn:243]: nit: double period at the end of this line should be a single period

**Review by andrewsavage1 — APPROVED** [review]
Please give others a chance to look and comment before submitting

**Review by kaidokert — COMMENTED** [review]
If builds+tests are passing this lgtm.  I'm not sure what's the right way to get this merged is though - if we do a non-squash merge, the commit history won't be great, but that may be acceptable here.  

The other part is orchestrating the container updates where needed ( eg all internal CI )

**Review by kaidokert — APPROVED** [review]

File: starboard/raspi/shared/test_filters.py (hunk 7 above)
Comment (y4vor) [reply starboard/raspi/shared/test_filters.py]: remove it then?

File: third_party/llvm-project/libcxx/include/__availability [hunk 12]
```diff
@@ -153,20 +153,23 @@
 
     // This controls the availability of floating-point std::to_chars functions.
     // These overloads were added later than the integer overloads.
-// #   define _LIBCPP_AVAILABILITY_HAS_NO_TO_CHARS_FLOATING_POINT
+    // Set to the same value as chromium upstream repo.
+#   define _LIBCPP_AVAILABILITY_HAS_NO_TO_CHARS_FLOATING_POINT 0
```
Comment (y4vor) [inline third_party/llvm-project/libcxx/include/__availability:157]: why do need these changes?

File: docker/linux/android/Dockerfile [hunk 13]
```diff
@@ -52,6 +52,7 @@ ARG CMDLINE_URL=https://dl.google.com/android/repository/${TOOLS}
 ENV ANDROID_SDK_ROOT=/root/starboard-toolchains/AndroidSdk/
 
 RUN cd /tmp \
+  && rm -rf ${ANDROID_SDK_ROOT} \
```
Comment (y4vor) [inline docker/linux/android/Dockerfile:55]: This change can be independently submitted, right?
Why is it part of the m114 upgrade?

File: third_party/llvm-project/libcxx/include/__availability (hunk 12 above)
Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/include/__availability]: These are set to upstream chromium llvm repo (see comment above (line156)

I get a failure:
[elf_loader_sand/1883:0315/184925.186056:FATAL:exported_symbols.cc(491)] Check failed: address. Failed to retrieve the address of '_ZNSt21_LIBCPP_ABI_NAMESPACE22__libcpp_verbose_abortEPKcz'.
 

https://github.com/youtube/cobalt/actions/runs/8300673760/job/22719946704?pr=2484

in evergreen test runs : so seems like this change is needed

File: docker/linux/android/Dockerfile (hunk 13 above)
Comment (niranjanyardi) [reply docker/linux/android/Dockerfile:55]: This fixes a failing test , idk why it's failing on this branch although it's not needed.
If you really want it , i can add it in a separate branch and submit it to trunk and it'll go away as part of the merge conflict resolution.  Brian has resolved merge conflicts once and so have I in this feature branch repo

File: starboard/raspi/shared/test_filters.py (hunk 7 above)
Comment (niranjanyardi) [reply starboard/raspi/shared/test_filters.py]: this will be resolved as part of the merge conflict resolution.
I want this to be here , otherwise it adds  noise due to failing tests ( cliock thorugh look at failures, many times unrelated to what you are doing).

Brian and I have already resolved merge conflicts twice, i just want the changes to be reviewed and all tests to pass before doing any more merge conflict resolutions

**Review by y4vor — APPROVED** [review]

File: third_party/llvm-project/libcxx/include/__availability (hunk 12 above)
Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/include/__availability]: reopening , this commit is getting us away from chromium upstream. I was looking at the wrong reference files(third_party/libc++ instead of buildtools/third_party/libc++

Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/include/__availability]: I ended up adding src/verbose_abort to the build source files - that fixed the error

**Review by briantting — APPROVED** [review]

File: third_party/llvm-project/compiler-rt/METADATA [hunk 14]
```diff
@@ -1,21 +1,26 @@
...
-  "Filtered subtree at third_party/llvm-project/compiler-rt."
+  "Subdirectory subtree of llvm-project."
 
 third_party {
-  url {
-    type: LOCAL_SOURCE
-    value: "https://cobalt.googlesource.com/third_party/llvm-project/compiler-rt_filtered_mirror"
+  identifier {
+    type: "ChromiumVersion"
+    value: "114.0.5735.351"
```
Comment (dahlstrom-g) [inline third_party/llvm-project/compiler-rt/METADATA:8]: This seems wrong—
Chromium doesn't include this tree at all, does it?

**Review by dahlstrom-g — CHANGES_REQUESTED** [review]

File: third_party/llvm-project/libcxx/METADATA [hunk 15]
```diff
@@ -1,21 +1,26 @@
...
   identifier {
     type: "ChromiumVersion"
-    value: "111.0.5563.150"  # from https://chromereleases.googleblog.com/2023/03/stable-channel-update-for-chromeos_29.html
+    value: "114.0.5735.351"
   }
   identifier {
     type: "Git"
-    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project/libcxx.git"
-    version: "1127c78cf90cf253be614a1e1d3645da57edbeb4"
-    # from https://chromium.googlesource.com/chromium/src/+/111.0.5563.150/DEPS#486
```
[the remark this answers is not in the document]
Comment (briantting) [reply third_party/llvm-project/libcxx/METADATA:14]: These versions are tied to the Chrome used Clang version they pinned for m114 (https://b.corp.google.com/issues/323184809#comment2)

File: third_party/llvm-project/compiler-rt/METADATA (hunk 14 above)
Comment (niranjanyardi) [reply third_party/llvm-project/compiler-rt/METADATA]: updated

File: third_party/llvm-project/libcxx/METADATA (hunk 15 above)
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/METADATA:14]: talked offline , these changes can be done in a follow up ( for changing METADATA for libcxx, libcxxabi, libunwind

**Review by briantting — APPROVED** [review]

File: third_party/llvm-project/libcxxabi/METADATA [hunk 16]
```diff
@@ -1,21 +1,26 @@
...
+    value: "114.0.5735.351"
   }
   identifier {
     type: "Git"
-    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project/libcxxabi.git"
-    version: "d520d582aa710cc0a4635620c02c5dbc187deb4f"
-    # from https://chromium.googlesource.com/chromium/src/+/111.0.5563.150/DEPS#466
+    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project"
+    version: "27f27d15f6c90b026eca23b8ee238fdbf772fd80"
+    # from https://chromium.googlesource.com/chromium/src/+/refs/tags/114.0.5735.351/tools/clang/scripts/update.py#38
```
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply third_party/llvm-project/libcxxabi/METADATA:14]: talked offline , investigating if compiler-rt is necessary . based on that will take next steps of whether to import this from chromium or llvm directly

File: third_party/llvm-project/libunwind/METADATA [hunk 17]
```diff
@@ -1,21 +1,26 @@
...
+    value: "114.0.5735.351"
   }
   identifier {
     type: "Git"
-    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project/libunwind.git"
-    version: "e95b94b74d26f8387d4fb03a687a2fab0ed8e91c"
-    # from https://chromium.googlesource.com/chromium/src/+/111.0.5563.150/DEPS#470
+    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project"
+    version: "27f27d15f6c90b026eca23b8ee238fdbf772fd80"
+    # from https://chromium.googlesource.com/chromium/src/+/refs/tags/114.0.5735.351/tools/clang/scripts/update.py#38
```
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply third_party/llvm-project/libunwind/METADATA:14]: talked offline , looking at whether compiler-rt is necessary, based on that will decide if we should be importing llvm from chrome or llvm directly

File: third_party/llvm-project/libcxx/METADATA [hunk 18]
```diff
@@ -1,21 +1,26 @@
...
+    value: "https://chromium.googlesource.com/external/github.com/llvm/llvm-project"
+    version: "27f27d15f6c90b026eca23b8ee238fdbf772fd80"
+    # from https://chromium.googlesource.com/chromium/src/+/refs/tags/114.0.5735.351/tools/clang/scripts/update.py#38
```
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply third_party/llvm-project/libcxx/METADATA:14]: talked offline , looking at whether compiler-rt is necessary, based on that will decide if we should be importing llvm from chrome or llvm directly

**Review by y4vor — APPROVED** [review]
